### PR TITLE
Decline unsubscribe related command in non-subscribed mode

### PR DESCRIPTION
**Pull request by hwware**
Now, when clients run the unsubscribe, sunsubscribe and punsubscribe commands in the non-subscribed mode, it returns 0.
Indeed this is a bug, we should not allow client run these kind of commands here.

Thus, this PR fixes this bug, but it is a break change for existing clients

## Update
This was introduced in Valkey8, and it is a breaking change, this will be revert in #1265

### CODE REVIEW

File: src/server.c [hunk 1]
```diff
@@ -3979,6 +3979,12 @@ int processCommand(client *c) {
         return C_OK;
     }
 
+    /* Not allow several UBSUBSCRIBE commands executed under non-pubsub mode */
+    if (!c->flag.pubsub && (c->cmd->proc == unsubscribeCommand || c->cmd->proc == sunsubscribeCommand ||
+                            c->cmd->proc == punsubscribeCommand)) {
+        rejectCommandFormat(c, "Can't execute '%s' command in non-subscribe mode", c->cmd->fullname);
```
Comment (zuiderkwast) [inline src/server.c]: Is this `-ERR`? Can we consider another prefix like `-NOSUB` and a more concise message like `-NOSUB Not in subscribed mode`?

There's no need to include the command name in the reply to the command. We don't do that in other errors.

File: tests/unit/info.tcl [hunk 2]
```diff
@@ -410,7 +410,8 @@ start_server {tags {"info" "external:skip"}} {
             set info [r info clients]
             assert_equal [getInfoProperty $info pubsub_clients] {1}
             # non-pubsub clients should not be involved
-            assert_equal {0} [unsubscribe $rd2 {non-exist-chan}]
+            catch {unsubscribe $rd2 {non-exist-chan}} e
+            assert_match {*Can't execute 'unsubscribe' command in non-subscribe mode*} $e
```
Comment (zuiderkwast) [inline tests/unit/info.tcl]: I prefer checking the uppercase error code like ERR or NOSUB in the test rather than the error message body.

**Review by zuiderkwast — COMMENTED** [review]
Great. We can discuss the exact error message.

I'm not sure we can call it a bug, but it is at least unexpected behaviour when called with more than one channel. I prefer we decribe it as changing an unexpected behaviour rather than a bugfix.

File: src/server.c [hunk 3]
```diff
@@ -3979,6 +3979,12 @@ int processCommand(client *c) {
         return C_OK;
     }
 
+    /* Not allow several UBSUBSCRIBE commands executed under non-pubsub mode */
```
Comment (hpatro) [inline src/server.c]:
```suggestion
    /* Not allow several UNSUBSCRIBE commands executed under non-pubsub mode */
```

File: src/server.c (hunk 1 above)
Comment (hpatro) [reply src/server.c]: Is it trivial for client(s) to handle new error code or should we just stick to `-ERR`? Don't recall many sub error opcode(s).

File: tests/unit/pubsub.tcl [hunk 4]
```diff
@@ -235,15 +235,17 @@ start_server {tags {"pubsub network"}} {
...
+        set rd1 [valkey_deferring_client]
+        assert_equal {1} [subscribe $rd1 {foo}]
+        assert_equal {2} [psubscribe $rd1 {foo.*}]
+        set reply1 [punsubscribe $rd1]
+        set reply2 [unsubscribe $rd1]
         concat $reply1 $reply2
-    } {punsubscribe {} 0 unsubscribe {} 0}
+        # clean up clients
+        $rd1 close
+    } {0}
```
Comment (hpatro) [inline tests/unit/pubsub.tcl]: You could move the cleanup one line and keep the concat as the last statement to verify the output.

File: tests/unit/pubsub.tcl [hunk 5]
```diff
@@ -105,8 +105,8 @@ start_server {tags {"pubsub network"}} {
 
     test "UNSUBSCRIBE from non-subscribed channels" {
         set rd1 [valkey_deferring_client]
-        assert_equal {0 0 0} [unsubscribe $rd1 {foo bar quux}]
-
+        catch {unsubscribe $rd1 {foo bar quux}} e
+        assert_match {*Can't execute 'unsubscribe' command in non-subscribe mode*} $e
         # clean up clients
         $rd1 close
     }
```
Comment (hpatro) [inline tests/unit/pubsub.tcl]: Maybe loop through all three [p|s]unsubscribe command in a single test?

File: src/server.c (hunk 1 above)
Comment (zuiderkwast) [reply src/server.c]: Yes, it's trivial for clients to handle a new code. Any error that's doesn't have a special handling is handled like any -ERR.

A separate code makes it possible for a client to distinguish this situation from any random -ERR. The clients should not match the error message text.

**Comment by hwware** [comment]
> Great. We can discuss the exact error message.
> 
> I'm not sure we can call it a bug, but it is at least unexpected behaviour when called with more than one channel. I prefer we decribe it as changing an unexpected behaviour rather than a bugfix.

Whatever it is a bug or unexpected behavior, this is a design error. Thus we need to fix it and make the right turn.

File: src/server.c (hunk 1 above)
Comment (hwware) [reply src/server.c]: Update the error as "NOSUB" prefix

File: tests/unit/pubsub.tcl (hunk 5 above)
Comment (hwware) [reply tests/unit/pubsub.tcl]: I combine the commands unsubscribe and punsubscribe in for loop, but for sunsubscribe  command,  I keep it in the pubsubshard.tcl file. Please take a look

**Review by zuiderkwast — APPROVED** [review]

**Comment by zuiderkwast** [comment]
This change introduced a race condition related to slot migration of sharded channel and at the same time client sending sunsubscribe, which can lead to the client getting out of sync in matching each reply to each command sent.

When a slot is migrated, the shard-channels are automatically usubscribed and clients will receive a spontaneous sunsubscribed message (push in RESP3) for each channel in the migrated slot. If the client also manually unsubscribes at the same time, the client can believe the spontaneous sunsubscribed response is a response to the SUNSUBSCRIBE command, but later it will receive an extra -NOSUB error after this PR. The client will believe this error is the response to the next command sent, such as `GET foo`, the -NOSUB error is matched with `GET foo` and returned as the reply to GET. Now the client's mapping of command-to-response is out of sync.

Scenario diagram:

```
Client                     Primary A       Primary B
  |                          |                |
  |  SUNSUBSCRIBE ch         |                |
  |------------------------->|                |
  |                        +-------------------------+
  |                        |  Slot migration A to B  |
  | [sunsubscribed, ch, 0] +-------------------------+
  |<-------------------------|                |
  |                          |                |
  | Error...                 |                |
  |<-------------------------|                |
  |                          |                |
```

An error reply is always in-band, i.e. it is always for a command.

If a client ignores errors for a command like subscribe and unsubscribe, then these commands get no in-band replies and one or more push replies. But there is an error, then subscribe/unsibscribe commands receives an in-band reply. Therefore, a correct client which sends `[S|P][UN]SUBSCRIBE` it must wait for either one push reply to confirm 'subscribed' or 'unsubscribed' or an error, before it can know that the command has been handled. If there are multiple channels, the remaining messages can just be handled as out-of-band.

But the above scenario with -NOSUB error breaks this logic. Now, there is no way for a client to make sure it stays in sync.

Therefore, I think we should maybe revert this PR. Without this PR, the client will instead in this scenario receive two 'sunsubscribed' push messages. Two push messages are easy to handle without getting out of sync.

**Comment by bjosv** [comment]
As stated before this breaks clients, one example is `hiredis` when running with the async API.
In async, hiredis waits for a successful reply to know the servers state and how to handle reply callbacks.
The handling in hiredis is a bit flawed from start, but now it would need to determine the result of an unsubscribe to avoid getting out of sync.

**Comment by zuiderkwast** [comment]
I made a test case to verify this. Actually, the corner case exists since before, ever since we added the sponaneous sunsubscribe message (when sharded pubsub was introduced?). After moving a slot and the client has received a sunsubscribe message, the SUNSUBSCRIBE command doesn't get a NOSUB error but a MOVED error, or a CLUSTERDOWN if the slot was deleted.

I simulated it with the following scenario, where Client 1 can be assumed to not have read the first message when it sends the SUNSUBSCRIBE command. Client 1 then believes that the `[sunsubscribe, ch, 0]` push message is received as a response to SUNSUBSCRIBE. The CLUSTERDOWN error reply remains to be read and it appears to be out of sync, i.e. to the client it doesn't appear to match a command it has sent. (If the client has sent another command in the pipeline, the CLUSTERDOWN appears to be its reply.)

```
Client 1                Client 2            Primary
  |                          |  DELSLOT        |
  |                          |---------------->|
  |                          |      OK         |
  | [sunsubscribe, ch, 0]    |<----------------|
  |<-------------------------------------------|
  |                          |                 |
  | SUNSUBSCRIBE ch          |                 |
  |------------------------------------------->|
  | -CLUSTERDOWN             |                 |
  |<-------------------------------------------|
  |                          |                 |
```

I don't see any way for a client to protect itself against this scenario. @bjosv do you?

**Comment by bjosv** [comment]
No, I don't think there is a way to protect against this currently..
